File: Problems/3Sum.py
# Gets the sum of 3 number through 2 pointers looping across sorted list
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()
        for i in range(len(nums)):
            if i > 0 and nums[i] == nums[i-1]:
                i += 1

            j = i + 1
            k = len(nums) - 1
            while j < k:
                add = nums[i] + nums[j] + nums[k]
                if add > 0:
                    k -= 1
                elif add < 0:
                    j += 1
                else:
                    res.append([nums[i], nums[j], nums[k]])
                    j += 1
                    while nums[j] == nums[j-1] and j < k:
                        j += 1
            
            return res
                    

File: Problems/3Sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        Arrays.sort(nums); // sort array to match sum
      
        for (int i = 0; i < nums.length; i++) { // loop through each number for possiuble sums
            if (i > 0 && nums[i] == nums[i-1]) { // skip repeating numbers
                continue;
            }
            
            int j = i + 1; 
            int k = nums.length - 1;

            while (j < k) { // find all sums between i and k
                int total = nums[i] + nums[j] + nums[k];

                if (total > 0) {
                    k--;
                } else if (total < 0) {
                    j++;
                } else {
                    res.add(Arrays.asList(nums[i], nums[j], nums[k])); // add if sum is found
                    j++;

                    while (nums[j] == nums[j-1] && j < k) {
                        j++;
                    }
                }
            }
        }
        return res;        
    }
}
